exercise-book/src/iterators.md

```diff
@@ -0,0 +1,359 @@
+
+## Step-by-Step-Solution
+
+In general, we also recommend using the Rust documentation to get unstuck. In particular, look for the examples in the [Iterator](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html) page of the standard library for this exercise.
+
+If you ever feel completely stuck or that you haven’t understood something, please hail the trainers quickly.
+
+### Step 1: New Project
+
+Create a new binary Cargo project and run it.
+
+Alternatively, use the [exercise-templates/iterators](../../exercise-templates/iterators/) template to get started.
+<details>
+  <summary>Solution</summary>
+
+```shell
+cargo new iterators
+cd iterators
+cargo run
+
+# if in exercise-book/exercise-templates/iterators
+cargo run --bin iterators1
+```
+
+Place the string
+
+```text
+//ignore everything that is not a number
+1
+2
+3
+4
+five
+6
+7
+∞
+9
+X
+```
+
+and place it in `iterators/numbers.txt`.
+</details>
+
+### Step 2: Read the string data
+
+Read the string data from a file placed in `iterators/numbers.txt`.
+Use the `reader.lines()` method to get rid of the newline characters.
+Collect it into a string with `.collect::<String>()` and print it to verify you're ingesting it correctly. It should have no newline characters since `lines()` trimmed them off.
+
+<details>
+  <summary>Solution</summary>
+
+We'll get rid of the `.unwrap()` in the next section.
+
+```rust
+#![allow(unused_imports)]
+use std::io::{BufRead, BufReader};
+use std::fs::File;
+use std::error::Error;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    use crate::*;
+    let f = File::open("../exercise-templates/iterators/numbers.txt")?;
+    let reader = BufReader::new(f);
+
+    let file_lines = reader.lines()
+        .map(|l| l.unwrap())
+        .collect::<String>();
+
+    println!("{:?}", file_lines);
+
+    Ok(())
+}
+```
+
+</details>
+
+### Step 3: Skip the non-numeric lines
+
+We'll collect into a `Vec<String>`s with [.parse()](https://doc.rust-lang.org/stable/std/primitive.str.html#method.parse) to show this intermediate step.
+
+Note that you may or may not need type annotations on `.parse()` depending on if you add them on the binding or not - that is, `let numeric_lines: Vec<i32> = ...` will give Rust type information to deduce the iterator's type correctly.
+
+<details>
+  <summary>Solution</summary>
+
+If the use of `filter_map` here is unfamiliar, go back and reread the ``Dealing with .unwrap()s in iterator chains`` section.
+
+```rust
+#![allow(unused_imports)]
+use std::io::{BufRead, BufReader};
+use std::fs::File;
+use std::error::Error;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    use crate::*;
+    let f = File::open("../exercise-templates/iterators/numbers.txt")?;
+    let reader = BufReader::new(f);
+
+    let numeric_lines = reader.lines()
+        .filter_map(|line| line.ok())
+        .filter_map(|line| line.parse::<i32>().ok())
+        .map(|stringy_num| stringy_num.to_string())
```

Comment:
Why are we converting to string here?

Reply:
Yeah this example can just be simple with the `Vec<i32>`. 

Thanks.